### PR TITLE
sdf/camera.sdf: fields for projection matrix

### PR DESCRIPTION
**Pull request by ihasdapie**
# 🎉 New feature

## Summary
This PR adds sdf fields to allow for configuration of the camera projection matrix.

Related PR in ign-sensors3:  https://github.com/gazebosim/gz-sensors/pull/249

## Test it
Make a camera with sdf tags specifying these new fields.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

### CODE REVIEW

**Review by chapulina — DISMISSED** [review]
Thank you for the PR! The C++ classes in SDFormat should reflect the XML spec. Do you mind adding the new elements here, with their respective tests?

https://github.com/gazebosim/sdformat/blob/3dd02ccb38351c2a115574d087cefd8764c2c3fa/sdf/1.9/camera.sdf#L175-L193

Also, note that we use `CamelCase` for function names, so the function names should be updated not to use `_p`. Maybe `SetLensProjection*`?

**Comment by ihasdapie** [comment]
Thanks for the feedback! I'll make those changes

**Review by ahcorde — DISMISSED** [review]
Do you mind to add tests for the new methods ?

File: sdf/1.7/camera.sdf [hunk 1]
```diff
@@ -151,6 +151,24 @@
       <element name="cy" type="double" default="120" required="1">
         <description>Y principal point (in pixels)</description>
       </element>
+      <element name="p_fx" type="double" default="277" required="0">
```
Comment (chapulina) [inline sdf/1.7/camera.sdf:154]: Thank you for adding the elements! I'm wondering if it's best to put these inside a new `<projection>` element that's a sibling of `intrinsics`. Or maybe a child of `intrinsics`? I'd have to refresh my computer vision knowledge.

Comment (scpeters) [reply sdf/1.7/camera.sdf:154]: yeah, I don't know enough about these parameters to have a useful opinion

**Comment by ihasdapie** [comment]
> Do you mind to add tests for the new methods ?
Yup, I missed it earlier. They've been added in https://github.com/gazebosim/sdformat/pull/1088/commits/a26db0ea77a6a4b184458eebaae6048363c9dfbe

File: sdf/1.7/camera.sdf (hunk 1 above)
Comment (ihasdapie) [reply sdf/1.7/camera.sdf:161]: > Thank you for adding the elements! I'm wondering if it's best to put these inside a new `<projection>` element that's a sibling of `intrinsics`. Or maybe a child of `intrinsics`? I'd have to refresh my computer vision knowledge.

IIUC the projection matrix is the cumulative effect of intrinsic and extrinsic parameters. So unless a xml element can have two parents (which I don't think is possible) I think @chapulina's suggestion to put these inside a new sibling `<projection>` element is best. I've done this in https://github.com/gazebosim/sdformat/commit/de4e5362bcd75140af26e1d9c2e43a7a11d5c2c8

References
- https://courses.cs.washington.edu/courses/cse576/17sp/notes/CamerasStereo17.pdf
- https://en.wikipedia.org/wiki/Camera_matrix

**Comment by scpeters** [comment]
> > Do you mind to add tests for the new methods ?
> 
> Yup, I missed it earlier. They've been added in [a26db0e](https://github.com/gazebosim/sdformat/commit/a26db0ea77a6a4b184458eebaae6048363c9dfbe)

to improve test coverage, let's add these new fields to the `sensors.sdf` test file and add corresponding test expectations to `link_dom.cc`

* https://github.com/ihasdapie/sdformat/blob/brianc/sdf9/configure_proj_mat/test/sdf/sensors.sdf#L23-L78
* https://github.com/ihasdapie/sdformat/blob/brianc/sdf9/configure_proj_mat/test/integration/link_dom.cc#L261-L304

**Comment by ihasdapie** [comment]
> to improve test coverage, let's add these new fields to the `sensors.sdf` test file and add corresponding test expectations to `link_dom.cc`

@scpeters  I've done so in https://github.com/gazebosim/sdformat/pull/1088/commits/47bb73a8e2b68a597c3c4ea941f886b71aa9e6de.

**Review by scpeters — APPROVED** [review]

**Comment by scpeters** [comment]
@ahcorde @chapulina you had requested changes, but I believe they have been addressed, so I dismissed your reviews. I'll merge on Monday unless I hear otherwise

**Pull request merged**